drivers/usb/udc/udc_mcux_ehci.c

```diff
@@ -0,0 +1,918 @@
+	struct udc_data *data = dev->data;
+	struct udc_mcux_data *priv = data->priv;
+	int err;
+
+	udc_mcux_get_hal_driver_id(priv, config);
+	if (priv->controller_id == 0xFFu) {
+		return -ENOMEM;
+	}
+
+	k_mutex_init(&data->mutex);
+
+	for (int i = 0; i < config->num_of_eps; i++) {
+		config->ep_cfg_out[i].caps.out = 1;
+		if (i == 0) {
+			config->ep_cfg_out[i].caps.control = 1;
+			config->ep_cfg_out[i].caps.mps = 64;
+		} else {
+			config->ep_cfg_out[i].caps.bulk = 1;
+			config->ep_cfg_out[i].caps.interrupt = 1;
+			config->ep_cfg_out[i].caps.iso = 1;
+			config->ep_cfg_out[i].caps.mps = 1024;
+		}
+
+		config->ep_cfg_out[i].addr = USB_EP_DIR_OUT | i;
+		err = udc_register_ep(dev, &config->ep_cfg_out[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	for (int i = 0; i < config->num_of_eps; i++) {
+		config->ep_cfg_in[i].caps.in = 1;
+		if (i == 0) {
+			config->ep_cfg_in[i].caps.control = 1;
+			config->ep_cfg_in[i].caps.mps = 64;
+		} else {
+			config->ep_cfg_in[i].caps.bulk = 1;
+			config->ep_cfg_in[i].caps.interrupt = 1;
+			config->ep_cfg_in[i].caps.iso = 1;
+			config->ep_cfg_in[i].caps.mps = 1024;
+		}
+
+		config->ep_cfg_in[i].addr = USB_EP_DIR_IN | i;
+		err = udc_register_ep(dev, &config->ep_cfg_in[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	data->caps.rwup = true;
+	data->caps.mps0 = USB_MCUX_MPS0;
+	data->caps.hs = true;
+	priv->dev = dev;
+
+	pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+
+	return 0;
+}
+
+static const struct udc_api udc_mcux_api = {
+	.device_speed = udc_mcux_device_speed,
+	.ep_enqueue = udc_mcux_ep_enqueue,
+	.ep_dequeue = udc_mcux_ep_dequeue,
+	.ep_set_halt = udc_mcux_ep_set_halt,
+	.ep_clear_halt = udc_mcux_ep_clear_halt,
+	.ep_try_config = NULL,
+	.ep_enable = udc_mcux_ep_enable,
+	.ep_disable = udc_mcux_ep_disable,
+	.host_wakeup = udc_mcux_host_wakeup,
+	.set_address = udc_mcux_set_address,
+	.enable = udc_mcux_enable,
+	.disable = udc_mcux_disable,
+	.init = udc_mcux_init,
+	.shutdown = udc_mcux_shutdown,
+	.lock = udc_mcux_lock,
+	.unlock = udc_mcux_unlock,
+};
+
+/* EHCI device driver interface */
+static const usb_device_controller_interface_struct_t udc_mcux_if = {
+	USB_DeviceEhciInit, USB_DeviceEhciDeinit, USB_DeviceEhciSend,
+	USB_DeviceEhciRecv, USB_DeviceEhciCancel, USB_DeviceEhciControl
+};
+
+#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED
+#define UDC_MCUX_PHY_CONFIGURE(n)							\
+	.need_to_init_phy = DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle) ? 1u : 0u,	\
+	.phy_config.D_CAL = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-d-cal, 0),	\
+	.phy_config.TXCAL45DP =								\
+		DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-cal-45-dp-ohms, 0),	\
+	.phy_config.TXCAL45DM =								\
+		DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-cal-45-dm-ohms, 0),
+#else
+#define UDC_MCUX_PHY_CONFIGURE(n)
+#endif
+
+#define USB_MCUX_EHCI_DEVICE_DEFINE(n)							\
```

Comment:
Now you can use `COND_CODE_1(DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle), ...) ` to define phy_config_##n and assign it to the pointer phy_config. I pushed all the changes to my branch, you can pull and squash it, and maybe adapt ip3511 driver accordingly. See https://github.com/jfischer-no/zephyr/commit/7fe0c43e6c1bce73bdded3ef073a4b7f026ad72b (https://github.com/jfischer-no/zephyr/commit/7fe0c43e6c1bce73bdded3ef073a4b7f026ad72b.patch)

Reply:
I have combine your changes with my side's codes because `phy_config` is not needed in the platforms that doesn't have the PHY. Please help to review.